lms/djangoapps/certificates/queue.py

```diff
@@ -306,6 +307,14 @@ def add_cert(self, student, course_id, course=None, forced_grade=None, template_
                 #   Despite blowing up the xml parser, bad values here are fine
                 grade_contents = None
 
+            # If this user's enrollment is not eligible to receive a
+            # certificate, mark it as such for reporting and
+            # analytics.
+            if not is_whitelisted and not CourseMode.is_eligible_for_certificate(enrollment_mode):
```

Comment:
Hrm, yeah, I mean this whole operation needs to be broken up into smaller pieces -- it's really a beast and is only going to get worse as more 'non-invasive' modifications are made to it.  So I guess it's a matter of how much you're willing to bite off right now as part of this PR.  

In terms of small-scale refactoring I might pull out the combination checks into separate code blocks that set boolean flags.  So for the first check here on 312 we might instead set a boolean flag named "enrollment_is_not_eligible" above and then the check can say "if enrollment_is_not_eligible".  Likewise for the grade_contents check below -- what are we actually checking on that line?  Could we restate it in a clearer way?

I understand if this is too much to deal with in the context of this PR -- just thought I'd see if we could make an incremental improvement or two.